Panic in compressor

Hello,

I'm opening this issue to report a panic which seems to originate in the lz4 code:
```
panic: runtime error: index out of range [65536] with length 65536

goroutine 611853 [running]:
github.com/pierrec/lz4/v4/internal/lz4block.(*Compressor).CompressBlock(0xc055bc0000, 0xc054c40000, 0x10000, 0x10000, 0xc05597e000, 0x10000, 0x10000, 0x70, 0x1071c60, 0xf32701)
        /go/src/github.com/segmentio/tracing/cmd/tracedb/vendor/github.com/pierrec/lz4/v4/internal/lz4block/block.go:191 +0xbbe
github.com/pierrec/lz4/v4/internal/lz4block.CompressBlock(0xc054c40000, 0x10000, 0x10000, 0xc05597e000, 0x10000, 0x10000, 0x7, 0xc03b56b588, 0xe2ca51)
        /go/src/github.com/segmentio/tracing/cmd/tracedb/vendor/github.com/pierrec/lz4/v4/internal/lz4block/block.go:91 +0xa7
github.com/pierrec/lz4/v4/internal/lz4stream.(*FrameDataBlock).Compress(0xc04d904150, 0xc0488fe240, 0xc054c40000, 0x10000, 0x10000, 0xc000000000, 0xe300f1)
        /go/src/github.com/segmentio/tracing/cmd/tracedb/vendor/github.com/pierrec/lz4/v4/internal/lz4stream/block.go:221 +0xa5
github.com/pierrec/lz4/v4.(*Writer).write(0xc0488fe2d0, 0xc054c40000, 0x10000, 0x10000, 0x200000000, 0xc03b56b638, 0xe2af4e)
        /go/src/github.com/segmentio/tracing/cmd/tracedb/vendor/github.com/pierrec/lz4/v4/writer.go:128 +0x7a
github.com/pierrec/lz4/v4.(*Writer).Write(0xc0488fe2d0, 0xc054c40000, 0x10000, 0x10000, 0x0, 0x0, 0x0)
        /go/src/github.com/segmentio/tracing/cmd/tracedb/vendor/github.com/pierrec/lz4/v4/writer.go:94 +0x508
```

For reference, this is how we constructed the writer:
```go
	x := lz4.NewWriter(w)
	x.Apply(
		lz4.BlockSizeOption(lz4.Block64Kb),
		lz4.ConcurrencyOption(1),
	)
```

In case this helps identify the issue, the software that hit this condition is writing buffers of exactly 64KB, and reusing the writers with calls to `Reset`.

This happened as well with the block size option set to 256KB, compressing data blocks of exactly 256 KB:

```
panic: runtime error: index out of range [262144] with length 262144

goroutine 25648 [running]:
github.com/pierrec/lz4/v4/internal/lz4block.(*Compressor).CompressBlock(0xc0b037e000, 0xc05473e000, 0x40000, 0x40000, 0xc0b03d0000, 0x40000, 0x40000, 0x101, 0xc064984da0, 0xc064984e18)
	/go/src/github.com/segmentio/tracing/cmd/tracedb/vendor/github.com/pierrec/lz4/v4/internal/lz4block/block.go:191 +0xbbe
github.com/pierrec/lz4/v4/internal/lz4block.CompressBlock(0xc05473e000, 0x40000, 0x40000, 0xc0b03d0000, 0x40000, 0x40000, 0x7, 0xc064984eb8, 0xe32071)
	/go/src/github.com/segmentio/tracing/cmd/tracedb/vendor/github.com/pierrec/lz4/v4/internal/lz4block/block.go:91 +0xa7
github.com/pierrec/lz4/v4/internal/lz4stream.(*FrameDataBlock).Compress(0xc051b07c00, 0xc051bea510, 0xc05473e000, 0x40000, 0x40000, 0xc000000000, 0xe35711)
	/go/src/github.com/segmentio/tracing/cmd/tracedb/vendor/github.com/pierrec/lz4/v4/internal/lz4stream/block.go:221 +0xa5
github.com/pierrec/lz4/v4.(*Writer).write(0xc051bea5a0, 0xc05473e000, 0x40000, 0x40000, 0x200000000, 0xc064984f68, 0xe3056e)
	/go/src/github.com/segmentio/tracing/cmd/tracedb/vendor/github.com/pierrec/lz4/v4/writer.go:128 +0x7a
github.com/pierrec/lz4/v4.(*Writer).Write(0xc051bea5a0, 0xc05473e000, 0x40000, 0x40000, 0x0, 0x0, 0x0)
```

--- a/fuzz/lz4.go
+++ b/fuzz/lz4.go
@@ -2,6 +2,7 @@ package lz4
 
 import (
 	"bytes"
+	"fmt"
 	"io"
 
 	"github.com/pierrec/lz4/v4"
@@ -41,6 +42,32 @@ func Fuzz(data []byte) int {
 		panic("not equal")
 	}
 	return 1
+}
+
+// CompressBlock with various destination sizes.
+// Shares its corpus with Fuzz.
+//
+// go-fuzz-build && go-fuzz -func=FuzzCompressBlock
+func FuzzCompressBlock(data []byte) int {
+	var (
+		bound = lz4.CompressBlockBound(len(data))
+		c     lz4.Compressor
+		comp  = make([]byte, lz4.CompressBlockBound(len(data)))
+		keep  = 0
+	)
+
+	for _, b := range []int{bound, len(data), len(data) - len(data)>>1} {
+		n, err := c.CompressBlock(data, comp[:b:b])
+
+		switch {
+		case err != nil && b == bound: // Should always work.
+			panic(err)
+		case n > b:
+			panic(fmt.Sprintf("n = %d > b = %d", n, b))
+		}
+	}
+
+	return keep
 }
 
 // Fuzzer for UncompressBlock: tries to decompress into a block the same size

--- a/internal/lz4block/block.go
+++ b/internal/lz4block/block.go
@@ -187,6 +187,9 @@ func (c *Compressor) CompressBlock(src, dst []byte) (int, error) {
 		}
 
 		mLen = si - mLen
+		if di >= len(dst) {
+			return 0, lz4errors.ErrInvalidSourceShortBuffer
+		}
 		if mLen < 0xF {
 			dst[di] = byte(mLen)
 		} else {

--- a/internal/lz4block/block_test.go
+++ b/internal/lz4block/block_test.go
@@ -10,6 +10,7 @@ import (
 
 	"github.com/pierrec/lz4/v4"
 	"github.com/pierrec/lz4/v4/internal/lz4block"
+	"github.com/pierrec/lz4/v4/internal/lz4errors"
 )
 
 type testcase struct {
@@ -162,5 +163,21 @@ func TestIssue23(t *testing.T) {
 		if got, want := n, 300; got > want {
 			t.Fatalf("not able to compress repeated data: got %d; want %d", got, want)
 		}
+	}
+}
+
+func TestIssue116(t *testing.T) {
+	src, err := ioutil.ReadFile("../../fuzz/corpus/pg1661.txt")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	dst := make([]byte, len(src)-len(src)>>1)
+	lz4block.CompressBlock(src, dst)
+
+	var c lz4block.Compressor
+	_, err = c.CompressBlock(src, dst)
+	if err != lz4errors.ErrInvalidSourceShortBuffer {
+		t.Fatalf("expected %v, got nil", lz4errors.ErrInvalidSourceShortBuffer)
 	}
 }

--- a/internal/lz4block/decode_test.go
+++ b/internal/lz4block/decode_test.go
@@ -126,10 +126,9 @@ func TestBlockDecode(t *testing.T) {
 		t.Run(test.name, func(t *testing.T) {
 			buf := make([]byte, len(test.exp))
 			n := decodeBlock(buf, test.src)
-			if n <= 0 {
-				t.Log(-n)
+			if n < 0 {
+				t.Log(n)
 			}
-
 			if !bytes.Equal(buf, test.exp) {
 				t.Fatalf("expected %q got %q", test.exp, buf)
 			}

